src/python/pants/engine/internals/graph.py

```diff
@@ -568,18 +568,81 @@ def __init__(
         )
 
 
+class InvalidFileDependency(Exception):
+    pass
+
+
+class ParsedDependencies(NamedTuple):
+    addresses: List[Address]
+    files: List[str]
+
+
+def parse_dependencies_field(
+    raw_value: Iterable[str], *, spec_path: str, subproject_roots: Sequence[str]
+) -> ParsedDependencies:
+    address_strings = []
+    files = []
+    for v in raw_value:
+        if ":" in v:
+            address_strings.append(v)
+            continue
+        elif v.startswith("./"):
+            files.append(PurePath(spec_path, v).as_posix())
+            continue
+
+        # Address specs for top-level targets take the form `//:foo`, meaning that we will have
+        # already recognized the `:`. Any other use of `//` is unnecessary and only used for
+        # clarity in BUILD files, so can be removed.
+        if v.startswith("//"):
+            v = v[2:]
+
+        if PurePath(v).suffix:
+            files.append(v)
+        else:
+            address_strings.append(v)
+    addresses = [
+        Address.parse(v, relative_to=spec_path, subproject_roots=subproject_roots)
+        for v in address_strings
+    ]
+    return ParsedDependencies(addresses, files)
+
+
+def validate_explicit_file_dep(address: Address, full_file: str, owners: Sequence[Address]) -> None:
+    if len(owners) > 1:
+        original_addresses = sorted(owner.maybe_convert_to_base_target().spec for owner in owners)
+        raise InvalidFileDependency(
+            f"The target {address} included {repr(full_file)} in its `dependencies` "
+            "field, but there are multiple owners of that file so it is ambiguous which one "
+            "Pants should use. Please instead change the `sources` fields of the owning "
+            "targets so that only one target owns this file, or choose which owner you want "
+            f"to use: {original_addresses}"
+        )
+    # If a file does not exist, but it matches the `sources` glob of a target, then it will
+    # have an owning target.
+    file_does_not_exist = len(owners) == 1 and not owners[0].generated_base_target_name
+    if not owners or file_does_not_exist:
+        raise InvalidFileDependency(
+            f"The target {address} included {repr(full_file)} in its `dependencies` "
+            "field, but there are no owners of that file. Please check that the file exists "
+            f"and that you spelled the file correctly."
```

Comment:
Should this also suggest declaring an owning target?

Reply:
Ah, good idea. I'll add that in the followup to add `!` ignore syntax.